lib/makara/proxy.rb

```diff
@@ -205,52 +210,57 @@ def appropriate_connection(method_name, args)
       end
     end
 
-    # master or replica
+    # primary or replica
     def appropriate_pool(method_name, args)
       # for testing purposes
       pool = _appropriate_pool(method_name, args)
       yield pool
     rescue ::Makara::Errors::AllConnectionsBlacklisted, ::Makara::Errors::NoConnectionsAvailable => e
-      if pool == @master_pool
-        @master_pool.connections.each(&:_makara_whitelist!)
+      if pool == @primary_pool
+        @primary_pool.connections.each(&:_makara_whitelist!)
         @replica_pool.connections.each(&:_makara_whitelist!)
         Kernel.raise e
       else
-        @master_pool.blacklist_errors << e
+        @primary_pool.blacklist_errors << e
         retry
       end
     end
 
     def _appropriate_pool(method_name, args)
-      # the args provided absolutely need master
-      if needs_master?(method_name, args)
-        stick_to_master(method_name, args)
-        @master_pool
+      # the args provided absolutely need primary
+      if needs_primary?(method_name, args)
+        stick_to_primary(method_name, args)
+        @primary_pool
 
-      elsif stuck_to_master?
+      elsif stuck_to_primary?
 
-        # we're on master because we already stuck this proxy in this
+        # we're on primary because we already stuck this proxy in this
         # request or because we got stuck in previous requests and the
         # stickiness is still valid
-        @master_pool
+        @primary_pool
 
       # all replicas are down (or empty)
       elsif @replica_pool.completely_blacklisted?
-        stick_to_master(method_name, args)
-        @master_pool
+        stick_to_primary(method_name, args)
+        @primary_pool
 
       elsif in_transaction?
-        @master_pool
+        @primary_pool
 
       # yay! use a replica
       else
         @replica_pool
       end
     end
 
-    # do these args require a master connection
-    def needs_master?(method_name, args)
-      true
+    # Do these args require a primary connection
+    def needs_primary?(method_name, args)
+      if respond_to?(:needs_master?)
+        warn "#{self.class}#needs_master? is deprecated. Switch to #needs_primary?"
+        needs_master?(method_name, args)
+      else
+        true
+      end
```

Comment:
I don't fully understand this change. How could it respond_to `needs_master?` since your removed that method? It also seems like it would be throwing a deprecation warning potentially when using the correct method. I'm sure there is some method magic happening that I just don't see but wanted to call it out.

Reply:
My understanding is this method is meant to be overwritten in subclasses by users. So if someone has defined their own `needs_master?`, we should still call it but emit this warning that they should rename the method.